### ISSUE TITLE
Sik-Ho Tang | Brief Review -- Residual Networks Behave Like Ensembles of Relatively Shallow Networks.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#111]
Sik-Ho Tang. [Brief Review — Residual Networks Behave Like Ensembles of Relatively Shallow Networks](https://sh-tsang.medium.com/brief-review-residual-networks-behave-like-ensembles-of-relatively-shallow-networks-5d9292796cbf).

**Comment by NorbertZheng** [comment]
## Overview
[Residual Networks Behave Like Ensembles of Relatively Shallow Networks](https://proceedings.neurips.cc/paper/2016/file/37bc2f75bf1bcfe8450a1a41c200364c-Reviews.html),
Veit NIPS’16, by Cornell University, 2016 NIPS, Over 1000 Citations. Image Classification.

 - Residual networks can be rewritten as an explicit collection of paths.
 - These paths show ensemble-like behavior in the sense that they do not strongly depend on each other.

**Comment by NorbertZheng** [comment]
## Residual Network Has Ensemble-Like Behavior
###  Preliminaries
Residual module in [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------) is popularly formulated as below:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/2ed20425-6cfe-488c-8774-f608334be1b4)

where $f_{i}()$ is the sequential operation of convolution ($W$), [batch norm ($B$)](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) and [ReLU](https://sh-tsang.medium.com/brief-review-rectified-linear-units-improve-restricted-boltzmann-machines-43724ede3ecf) ($sigma$):
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c2020e00-5304-40c0-867b-0aa37f674ca5)

[Highway](https://towardsdatascience.com/review-highway-networks-gating-function-to-highway-image-classification-5a33833797b5?source=post_page---------------------------) network indeed is a special case of [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------).
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/5a0e02b8-595a-49e5-a8f8-4eb7e02c5c0e)

(Please feel free to read [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------) and [Highway](https://towardsdatascience.com/review-highway-networks-gating-function-to-highway-image-classification-5a33833797b5?source=post_page---------------------------) if interested.)

**Comment by NorbertZheng** [comment]
### Ensemble-Like Behavior
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c00712af-2f33-41f9-93e9-4ba09fcb3b9e)
An unraveled view of a 3-block residual network.

 - The output of each stage is based on the combination of two sub-terms.

Thus, **the shared structure of the residual network apparent** by unrolling the recursion into an exponential number of nested terms, expanding one layer at each substitution step:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/cd981f79-e3e2-4a27-ae9c-fbf76a6d984a)

The graph makes clear that data flows along many paths from input to output. Each path is a unique configuration of which residual module to enter and which to skip.

**Comment by NorbertZheng** [comment]
### Comparison to Plain Network
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/80ab863e-8c4d-4b0f-b353-819932bfd11b)
Plain Network such as [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160?source=post_page---------------------------) and [VGGNet](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11?source=post_page---------------------------).

In the classical plain network, each layer of processing depends only on the output of the previous layer.

However, in residual networks, each module $f_{i}()$ in the residual network is fed data from a mixture of $2^{(i-1)}$ different distributions generated from every possible configuration of the previous $i-1$ residual modules.

**Comment by NorbertZheng** [comment]
## Analytic Results
### Deleting individual layers from neural networks at test time
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/5e7c3933-3ff7-4c6d-a1e9-87a6cff75c97)
Deleting a layer in residual networks at test time (a) is equivalent to zeroing half of the paths.

When a layer is removed, the number of paths is reduced from $2^{n}$ to $2^{(n-1)}$, leaving half the number of paths valid.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d17baed1-b0e5-4ba0-ae0b-175928adf014)
Figure 3: Deleting individual layers from [VGGNet](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11?source=post_page---------------------------)/[ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------) on Left: CIFAR-10, Right: ImageNet.

 - **Deleting any layer in [VGGNet](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11?source=post_page---------------------------) reduces performance to chance levels. Surprisingly, this is NOT the case for [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------).**

**Comment by NorbertZheng** [comment]
### Deleting/Reordering many modules from residual networks at test-time
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/3b33eaac-0fef-403a-ac01-940ab6c60c6a)
(a) Error increases smoothly when randomly deleting several modules from a residual network. (b) Error also increases smoothly when re-ordering a residual network by shuffling building blocks.

**One key characteristic of ensembles is that their performance depends smoothly on the number of members.**

 - (a): Deleting increasing numbers of residual modules, increases error smoothly. This implies residual networks behave like ensembles.
 - (b): $k$ randomly sampled pairs of building blocks with compatible dimensionality are swapped. As corruption increases, the error smoothly increases as well.

**Comment by NorbertZheng** [comment]
### The importance of short paths in residual networks
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e0840812-c13a-4267-8cbc-afe62ea348a4)
(a) The distribution of all possible path lengths. This follows a Binomial distribution. (b) How much gradient is induced on the first layer of the network through paths of varying length, which appears to decay roughly exponentially with the number of modules the gradient passes through. (c) multiply these two functions (a) & (b), see how much gradient comes from all paths of a certain length.

Surprisingly, almost all of the gradient updates during training come from paths between 5 and 17 modules long.

 - These are the effective paths, even though they constitute only 0.45% of all paths through this network. Moreover, in comparison to the total length of the network, **the effective paths are relatively shallow**.

**Comment by NorbertZheng** [comment]
### [Stochastic Depth](https://towardsdatascience.com/review-stochastic-depth-image-classification-a4e225807f4a?source=post_page---------------------------)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e1efe8a8-611c-4ad2-a451-2017b67aa443)
Left: Fraction of paths remaining after deleting individual layers. Right: Impact of [Stochastic Depth](https://towardsdatascience.com/review-stochastic-depth-image-classification-a4e225807f4a?source=post_page---------------------------) on resilience to layer deletion.

Removing residual modules mostly removes long paths.

 - Left: Even after deleting 10 residual modules, many of the effective paths between 5 and 17 modules long are still valid.
 - Right: A random subset of the residual modules is selected for each mini-batch during training. Training with [Stochastic Depth](https://towardsdatascience.com/review-stochastic-depth-image-classification-a4e225807f4a?source=post_page---------------------------) improves resilience slightly.

(Please feel free to read [Stochastic Depth](https://towardsdatascience.com/review-stochastic-depth-image-classification-a4e225807f4a?source=post_page---------------------------) if interested.)

**Comment by NorbertZheng** [comment]
Right now, there are many models proposed with much fewer number of weight layers, i.e. shallower.

**Comment by NorbertZheng** [comment]
## References
 - [2016 NeurIPS] [Veit] [Residual Networks Behave Like Ensembles of Relatively Shallow Networks](https://proceedings.neurips.cc/paper/2016/file/37bc2f75bf1bcfe8450a1a41c200364c-Reviews.html).
